pkg/spade-spike/spade-spike.go

```diff
@@ -0,0 +1,286 @@
+	return BitswapClient{
+		host:    host,
+		timeout: timeout,
+	}
+}
+
+// Attempts to decode the block data into a node and return its links
+func FindLinks(ctx context.Context, blk blocks.Block) ([]datamodel.Link, error) {
+	if blk.Cid().Prefix().Codec == goCid.Raw {
+		// This can happen at the bottom of the tree
+		return nil, errors.New("raw block encountered " + blk.Cid().String())
+	}
+
+	decoder, err := cidlink.DefaultLinkSystem().DecoderChooser(cidlink.Link{Cid: blk.Cid()})
+
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := ipld.Decode(blk.RawData(), decoder)
+	if err != nil {
+		return nil, err
+	}
+
+	links, err := traversal.SelectLinks(node)
+	if err != nil {
+		return nil, err
+	}
+
+	return links, nil
+}
+
+// Returns the raw block data, the links, and error if any
+func (c BitswapClient) RetrieveBlock(
+	parent context.Context,
+	target peer.AddrInfo,
+	cid goCid.Cid) (blocks.Block, error) {
+	network := bsnet.NewFromIpfsHost(c.host, SingleContentRouter{
+		AddrInfo: target,
+	})
+	bswap := bsclient.New(parent, network, blockstore.NewBlockstore(datastore.NewMapDatastore()))
+	notFound := make(chan struct{})
+	network.Start(MessageReceiver{BSClient: bswap, MessageHandler: func(
+		ctx context.Context, sender peer.ID, incoming bsmsg.BitSwapMessage) {
+		if sender == target.ID && slices.Contains(incoming.DontHaves(), cid) {
+			logger.Info("Block not found")
+			close(notFound)
+		}
+	}})
+	defer bswap.Close()
+	defer network.Stop()
+	connectContext, cancel := context.WithTimeout(parent, c.timeout)
+	defer cancel()
+	logger.Info("Connecting to target peer...")
+	err := c.host.Connect(connectContext, target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to target peer, %s", err)
+	}
+
+	resultChan := make(chan blocks.Block)
+	errChan := make(chan error)
+	go func() {
+		logger.Debug("Retrieving block...")
+		blk, err := bswap.GetBlock(connectContext, cid)
+		if err != nil {
+			logger.Info(err)
+			errChan <- err
+		} else {
+			resultChan <- blk
+		}
+	}()
+	select {
+	case <-notFound:
+		return nil, errors.New("DONT_HAVE received from the target peer")
+
+	case blk := <-resultChan:
+		return blk, nil
+
+	case err := <-errChan:
+		return nil, fmt.Errorf("error received %s", err)
+	}
+}
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes
+// Returns true if all retrievals were successful, false if any failed
+func SpadeTraversal(ctx context.Context, startingCid goCid.Cid, p peer.AddrInfo) (bool, error) {
+	cidToRetrieve := startingCid
+
+	// It will always be 3 layers, no more no less
+	// ref: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
+	for i := 0; i <= 3; i++ {
+		// For some reason, need to re-init the host & client every time we do a fetch
+		// otherwise, we get context timeout error after the first fetch
+		host, err := net.InitHost(ctx, nil)
```

Comment:
@xinaxu do you happen to know anything about this? IMO it would make sense to just init the host/Bitswap client just once, but when I do that I can only successfully retrieve the first CID, all subsequent ones get context timeout, no matter how long I set the timeout limit

Reply:
Not sure, I need to see your code that reuses the same host